packages/frontend/core/src/modules/cmdk/entities/quick-search.ts

```diff
@@ -0,0 +1,145 @@
+import type {
+  DocRecord,
+  DocsService,
+  WorkspaceService,
+} from '@toeverything/infra';
+import { Entity, LiveData } from '@toeverything/infra';
+
+import { resolveLinkToDoc } from '../../navigation';
+
+type QuickSearchMode = 'commands' | 'docs';
+
+export type SearchCallbackResult =
+  | {
+      docId: string;
+      blockId?: string;
+    }
+  | {
+      query: string;
+      action: 'insert';
+    };
+
+// todo: move command registry to entity as well
+export class QuickSearch extends Entity {
+  constructor(
+    private readonly docsService: DocsService,
+    private readonly workspaceService: WorkspaceService
+  ) {
+    super();
+  }
+  private readonly state$ = new LiveData<{
+    mode: QuickSearchMode;
+    query: string;
+    callback?: (result: SearchCallbackResult) => void;
+  } | null>(null);
+
+  readonly show$ = this.state$.map(s => !!s);
+
+  show = (
+    mode: QuickSearchMode | null = 'commands',
+    opts: {
+      callback?: (res: SearchCallbackResult) => void;
+      query?: string;
+    } = {}
+  ) => {
+    if (mode === null) {
+      this.state$.next(null);
+    } else {
+      this.state$.next({
```

Comment:
```suggestion
      if (this.state$.value) {
        this.state$.value.callback(null)
      }
      this.state$.next({
```

Reply:
Do you mean change `show` to a toggle function? I think could be cases that quick search brought up again even when it is active.

Reply:
It seems that when opening the next quicksearch, the previous one is not closed, callback(null) is not called anywhere.

Reply:
Oh I see what you mean.